### ISSUE TITLE
Crashing of pfp

### ISSUE THREAD

**Issue by Fuzion24** [d0c-s4vage/pfp#10]
```
➜  stagefright_media_files git:(master) ✗ cat blah.py
import pfp

dom = pfp.parse(
data_file="./cve-2015-3864.mp4",
template_file="~/Downloads/010templates/MP4Template.bt"
)
```

Using this template: https://github.com/nevil/010-Templates/blob/master/MP4Template.bt
And this .mp4 : https://google-security-research.googlecode.com/issues/attachment?aid=5020000000&name=subtitletest.mp4 from (https://code.google.com/p/google-security-research/issues/detail?id=502)

```
➜  stagefright_media_files git:(master) ✗ python blah.py
/var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:50:30: warning: missing terminating ' character [-Winvalid-pp-token]
    <!-- </textarea> --><!-- '"` --><meta content="authenticity_token" name="csrf-param" />
                             ^
/var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:95:32: warning: missing terminating ' character [-Winvalid-pp-token]
      <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/nevil/010-Templates/search" class="js-site-search-form" data-global-search-url="/search" data-rep...
                               ^
/var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:273:1244: warning: missing terminating '"' character [-Winvalid-pp-token]
  .../></div><button class="btn-link js-clone-selector" data-protocol="subversion" type="submit">Subversion</button></form>.
                                                                                               ^
/var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:388:131: warning: missing terminating '"' character [-Winvalid-pp-token]
            <a href="/nevil/010-Templates/commit/4224b84d03985055bc4e1fc2bee2d95d92803c6d" class="message" data-pjax="true" title="H265
                                                                                                                                  ^
/var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:391:21: warning: missing terminating '"' character [-Winvalid-pp-token]
Initial H265 support">H265</a>
                    ^
/var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:7281:28: warning: missing terminating ' character [-Winvalid-pp-token]
  <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="" class="js-jump-to-line-form" method="get"><div style="margin:0;padding:0;display:inline"><input name...
                           ^
6 warnings generated.
Traceback (most recent call last):
  File "blah.py", line 5, in <module>
    template_file="~/Downloads/010templates/MP4Template.bt"
  File "/usr/local/lib/python2.7/site-packages/pfp/__init__.py", line 62, in parse
    dom = interp.parse(data, template, predefines=predefines, orig_filename=orig_filename, keep_successful=keep_successful)
  File "/usr/local/lib/python2.7/site-packages/pfp/interp.py", line 573, in parse
    self._ast = self._parse_string(template, predefines)
  File "/usr/local/lib/python2.7/site-packages/pfp/interp.py", line 708, in _parse_string
    keep_scopes=predefines
  File "/usr/local/lib/python2.7/site-packages/py010parser/__init__.py", line 113, in parse_string
    return parser.parse(text, filename, predefine_types=predefine_types, keep_scopes=keep_scopes)
  File "/usr/local/lib/python2.7/site-packages/py010parser/c_parser.py", line 160, in parse
    debug=debuglevel)
  File "/usr/local/lib/python2.7/site-packages/py010parser/ply/yacc.py", line 265, in parse
    return self.parseopt_notrack(input,lexer,debug,tracking,tokenfunc)
  File "/usr/local/lib/python2.7/site-packages/py010parser/ply/yacc.py", line 1047, in parseopt_notrack
    tok = self.errorfunc(errtoken)
  File "/usr/local/lib/python2.7/site-packages/py010parser/c_parser.py", line 1931, in p_error
    column=self.clex.find_tok_column(p)))
  File "/usr/local/lib/python2.7/site-packages/py010parser/plyparser.py", line 54, in _parse_error
    raise ParseError("%s: %s" % (coord, msg))
py010parser.plyparser.ParseError: /var/folders/6h/h97znl5d0qb9pmpb6ldcbr1m0000gn/T/tmp1XLs3K:4:1: before: <
```

**Comment by d0c-s4vage** [comment]
It looks like you have pulled down the html of the github page for the `MP4Template.bt` instead of the raw content of the template from github. wget this url instead: `https://raw.githubusercontent.com/nevil/010-Templates/master/MP4Template.bt`

I tried running the mp4 template and filed issues #11 and #12 to fix some other errors I noticed.

**Comment by Fuzion24** [comment]
Oh, geez. That's really embarrassing on my part.

After downloading the proper template, the parser still seems to be crashing:

```
Traceback (most recent call last):
  File "blah.py", line 5, in <module>
    template_file="~/Downloads/010templates/MP4Template.bt"
  File "/usr/local/lib/python2.7/site-packages/pfp/__init__.py", line 62, in parse
    dom = interp.parse(data, template, predefines=predefines, orig_filename=orig_filename, keep_successful=keep_successful)
  File "/usr/local/lib/python2.7/site-packages/pfp/interp.py", line 573, in parse
    self._ast = self._parse_string(template, predefines)
  File "/usr/local/lib/python2.7/site-packages/pfp/interp.py", line 708, in _parse_string
    keep_scopes=predefines
  File "/usr/local/lib/python2.7/site-packages/py010parser/__init__.py", line 113, in parse_string
    return parser.parse(text, filename, predefine_types=predefine_types, keep_scopes=keep_scopes)
  File "/usr/local/lib/python2.7/site-packages/py010parser/c_parser.py", line 160, in parse
    debug=debuglevel)
  File "/usr/local/lib/python2.7/site-packages/py010parser/ply/yacc.py", line 265, in parse
    return self.parseopt_notrack(input,lexer,debug,tracking,tokenfunc)
  File "/usr/local/lib/python2.7/site-packages/py010parser/ply/yacc.py", line 971, in parseopt_notrack
    p.callable(pslice)
  File "/usr/local/lib/python2.7/site-packages/py010parser/c_parser.py", line 734, in p_decl_body_2
    raise ParseError("bitfields may only be used inside of structs")
py010parser.plyparser.ParseError: bitfields may only be used inside of structs
```

**Comment by d0c-s4vage** [comment]
NP :^) I've done that before too.

Ah. That issue was fixed in the latest version of pfp (do `pip install
--upgrade pfp`).

IIRC there is still one issue (#11) that I need to address before the
`MP4Template.bt` will work. I'll get to that tomorrow morning.

Thanks for giving pfp a spin!

**Comment by d0c-s4vage** [comment]
Fixed.

Besides the other two issues I created (#11 and #12), I also had to open d0c-s4vage/py010parser#5 to get this fixed.

I tested the MP4Template and compared the output to 010's parsing and everything matched up.

**Comment by Fuzion24** [comment]
which media file did you use to test?

After updating pfp, I'm currently seeing this:

```
template #: 0In 'trak' Skipping box: 'tx3g' close to offset 0xf0
In 'trak' Skipping box: 'fake' close to offset 0x64f8
In 'trak' Skipping box: 'tx3g' close to offset 0x6510
In 'trak' Skipping box: 'x3g
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
In 'trak' Skipping box: '<e0><e0><e0><e0>' close to offset 0x6685
```

**Comment by d0c-s4vage** [comment]
Ah, I see it now. Hehe this has brought out tons of bugs in pfp. I'll put in a fix at lunch for #13

**Comment by d0c-s4vage** [comment]
@Fuzion24 afaict it's working correctly now. If there's more issues I'll reopen the ticket. Changes are in v0.1.14
